plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/ui/contentAssist/ContentAssistFragment2.xtend

```diff
@@ -0,0 +1,330 @@
+		val processedTerminals = newHashSet();
+		
+		// for each type of terminal occurring in 'terminals' (evaluated lazily) ...
+		val candidates = terminals.fold(<AbstractElement>newHashSet) [ candidates, terminal |
+			if (!processedTerminals.contains(terminal.eClass)) {
+				processedTerminals += terminal.eClass
+				candidates += terminal
+			}
+			candidates
+		]
+		
+		// ... generate an 'instanceof' clause
+		'''
+			�FOR terminal : candidates�
+				if (assignment.getTerminal() instanceof �terminal.eClass.instanceClass�) {
+					�terminal.assignmentTerminal("assignment.getTerminal()")�
+				}
+			�ENDFOR�		
+		'''
+	}
+
+	private def dispatch StringConcatenationClient assignmentTerminal(AbstractElement element, String accessor) {
+		'''
+		// subclasses may override
+		'''
+	}	
+
+	private def dispatch StringConcatenationClient assignmentTerminal(CrossReference element, String accessor) {
+		'''
+		lookupCrossReference(((�CrossReference�)�accessor�), context, acceptor);
+		'''
+	}	
+
+	private def dispatch StringConcatenationClient assignmentTerminal(RuleCall element, String accessor) {
+		'''
+		completeRuleCall(((�RuleCall�)�accessor�), context, acceptor);
+		'''
+	}
+	
+	private def dispatch StringConcatenationClient assignmentTerminal(Alternatives alternatives, String accessor) {
+		val list = alternatives.elements
+		'''
+			�FOR it : list�
+				�it.assignmentTerminal("((Alternatives)" + accessor + ").getElements.get("+ list.indexOf(it) +")")�
+			�ENDFOR�
+		'''
+	}
+
+	// helper methods
+
+	private def getContentAssistContextClass() {
+		new TypeReference("org.eclipse.xtext.ui.editor.contentassist.ContentAssistContext")
+	}
+
+	private def getICompletionProposalAcceptorClass() {
+		new TypeReference("org.eclipse.xtext.ui.editor.contentassist.ICompletionProposalAcceptor")
+	} 
+
+	def private getFQFeatureName(AbstractRule r) {
+		"_" + r.name;
+	}
+
+	def private getFQFeatureName(Assignment a) {
+		a.containingParserRule().name.toFirstUpper() + "_" + a.feature.toFirstUpper();
+	}
+
+	def private computeFQFeatureNames(Grammar g) {
+		g.containedAssignments.map[ FQFeatureName ] + g.rules.map[ FQFeatureName ]
+	}
+
+	def getFQFeatureNamesToExclude(Grammar g) {
+		val superGrammar = g.nonTerminalsSuperGrammar
+		if (superGrammar != null) {
+			val superGrammarsFqFeatureNames = computeFQFeatureNamesFromSuperGrammars(g).toSet
+			val thisGrammarFqFeatureNames = g.computeFQFeatureNames.toSet
+
+			// return all elements that are already defined in some inherited grammars
+			Sets.intersection(thisGrammarFqFeatureNames, superGrammarsFqFeatureNames)
+		} else {
+			newHashSet()
+		}
+	}
+
+	def private computeFQFeatureNamesFromSuperGrammars(Grammar g) {
+		val superGrammars = newHashSet()
+		computeAllSuperGrammars(g, superGrammars)
+		superGrammars.map[
+			computeFQFeatureNames
+		].flatten
+	}
+
+	def private void computeAllSuperGrammars(Grammar current, Set<Grammar> visitedGrammars) {
+		for (s : current.usedGrammars) {
```

Comment:
thanks! didn't find it on my own (``computeAllSuperGrammars`` stems from the existing implementation)